File: src/network/packet/impl/PlayerTextPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl;

use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\data\TextType;
use pocketcloud\cloud\network\packet\util\PacketData;
use pocketcloud\cloud\player\CloudPlayerManager;

final class PlayerTextPacket extends CloudPacket implements CloudboundPacket, ClientboundPacket {

    public function __construct(
        private string $player = "",
        private string $text = "",
        private ?TextType $type = null
    ) {}

    public function handle(ServerClient $client): void {
        if (($player = CloudPlayerManager::getInstance()->get($this->player)) !== null) {
            $player->send($this->text, $this->type);
        }
    }

    public function encodePayload(PacketData $packetData): void {
        $packetData->writeAll($this->player, $this->text, $this->type);
    }

    public function decodePayload(PacketData $packetData): void {
        $packetData->readAllTypeSafe([&$this->player, &$this->text, &$this->type], [fn() => $packetData->readString(), fn() => $packetData->readString(), fn() => $packetData->readTextType()]);
    }

    public function getPlayer(): string {
        return $this->player;
    }

    public function getText(): string {
        return $this->text;
    }

    public function getType(): ?TextType {
        return $this->type;
    }

    public static function create(string $player, string $text, TextType $type): self {
        return new self($player, $text, $type);
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/impl/PlayerTextPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.ClientboundPacket;
import de.pocketcloud.cloud.network.packet.CloudboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.cloud.network.packet.type.TextType;
import de.pocketcloud.cloud.network.packet.data.PacketData;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

@NoArgsConstructor
@Getter
public final class PlayerTextPacket extends CloudPacket implements CloudboundPacket, ClientboundPacket {

    private String player;
    private String text;
    private TextType type;

    public PlayerTextPacket(String player, String text, TextType type) {
        this.player = player != null ? player : "";
        this.text = text != null ? text : "";
        this.type = type;
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        CloudPlayerManager.instance().get(player).ifPresent(p -> p.send(text, type));
    }

    @Override
    public void encodePayload(PacketData packetData) {
        packetData.writeAll(player, text, type);
    }

    @Override
    public void decodePayload(PacketData packetData) {
        this.player = packetData.readString();
        this.text = packetData.readString();
        this.type = packetData.readTextType();
    }

    public static PlayerTextPacket create(String player, String text, TextType type) {
        return new PlayerTextPacket(player, text, type);
    }
}
